### PR TITLE
feature/639404034-create-login-page

### PR DESCRIPTION
**Pull request by woutermont**

### CODE REVIEW

File: packages/semcom-demo/src/app/connect/services/provider.effects.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      .pipe(
+        map(providers => (ProvidersLoaded({ providers }))),
+        catchError(() => EMPTY)
+      ))
+    )
+  );
+
+  connectProvider$ = createEffect(() => this.actions$.pipe(
+    ofType(ProviderSelected),
+    tap(action => login({ oidcIssuer: action.provider.url, redirectUrl: 'http://localhost:4200/connect/callback' }))
```
Comment (wouteraj) [inline packages/semcom-demo/src/app/connect/services/provider.effects.ts:30]: - Unit testing would be more convenient if there's no business logic in an effect. Maybe move this functionality to a service?
 - It's best not to hardcode the base url. When hosting the app, or even serving on a different port would result in an incorrect `redirectUrl`.

File: packages/semcom-demo/src/app/connect/services/provider.service.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { Injectable } from '@angular/core';
+import { Observable } from 'rxjs';
+import { Provider } from '../models/provider.model';
+
+@Injectable({ providedIn: 'root' })
+export class ProviderService {
+
+  getAll(): Observable<Provider[]> {
+    return new Observable(subscriber => subscriber.next([
```
Comment (wouteraj) [inline packages/semcom-demo/src/app/connect/services/provider.service.ts:9]: You could use the `of` operator to simplify this a little.

https://www.learnrxjs.io/learn-rxjs/operators/creation/of

File: packages/semcom-demo/src/app/connect/connect.component.spec.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
+import { declarations, imports, providers } from '../app.module';
+import { ConnectComponent } from './connect.component';
+import { RouterTestingModule } from '@angular/router/testing';
+import { TestBed } from '@angular/core/testing';
+
+describe('HomeComponent', () => {
```
Comment (wouteraj) [inline packages/semcom-demo/src/app/connect/connect.component.spec.ts:6]: This looks like a small typo.

File: packages/semcom-demo/src/app/connect/connect.component.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    providers$: Observable<Provider[]> = this.store.select<Provider[]>(state => state.connectState.providers);
+
+    constructor(private store: Store<AppState>) {}
+
+    ngOnInit(): void {
+      this.store.dispatch(ConnectPageInit());
+    }
+
+    connect(provider: Provider): void {
```
Comment (wouteraj) [inline packages/semcom-demo/src/app/connect/connect.component.ts:23]: Don't forget to check for `null`.

File: packages/semcom-demo/src/app/connect/connect.guard.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+@Injectable({
+  providedIn: 'root',
+})
+export class ConnectGuard implements CanActivate, Resolve<void> {
+
+  sessionInfo$: Observable<ISessionInfo|null> = this.store.select(state => state.connectState.sessionInfo);
+
+  constructor(private store: Store<AppState>, private router: Router, private location: Location) {}
+
+  resolve(): void {
```
Comment (wouteraj) [inline packages/semcom-demo/src/app/connect/connect.guard.ts:20]: Very nice; great idea to implement this in a resolve function. 

Just a little suggestion: wouldn't it be clearer to define split this off in a separate class, given that a `Resolver` is not conceptually the same as a `Guard`?

**Review by wouteraj — CHANGES_REQUESTED** [review]
Look very good!! Some general questions/remarks:

 - I noticed that you defined actions and effects at the component-level (like `connect.actions.ts`), while defining your state and reducers at the application-level (like `app.reducers.ts`). Given the limited scope of the demo, wouldn't it be easier to define the actions and effects at the application-level as well?
 - You didn't write unit tests yet. This is obviously not relevant in the context of the demo, but might be useful in the context of your onboarding.

File: packages/semcom-demo/src/app/connect/connect.component.ts (hunk 4 above)
Comment (woutermont) [reply packages/semcom-demo/src/app/connect/connect.component.ts:23]: Aha, right ... I was under the impression that Typescript makes that unneccessary. That's not the case?

File: packages/semcom-demo/src/app/connect/connect.guard.ts (hunk 5 above)
Comment (woutermont) [reply packages/semcom-demo/src/app/connect/connect.guard.ts:20]: Actually, Angular docs describe it as a guard. That's wy I thought it nice to put all connect-related guard-logic in the same class.

**Comment by woutermont** [comment]
@wouteraj 

> * I noticed that you defined actions and effects at the component-level (like `connect.actions.ts`), while defining your state and reducers at the application-level (like `app.reducers.ts`). Given the limited scope of the demo, wouldn't it be easier to define the actions and effects at the application-level as well?

I think that, independent of the size of the project, the concepts of NgRx are more closely linked in 3 groups, which make it i.m.o. clearer. Effects seem inherently linked to services, and actions always originate either in these effects/services or in components. It makes therefore sense i.m.o. to put effects near their services, and actions near the effects/services or components where they typically originate. Reducers are inherently linked to states, and do not necessarily follow the hierarchy of components, and can therefore be loosely grouped around modules. Just my 2 cents.

**Comment by woutermont** [comment]
Also, about the unit tests: what would I test a.t.m.?

**Pull request merged**